### ISSUE TITLE
Background for a Robotics/AI engineer

### ISSUE THREAD

**Issue by trangml** [trangml/swyxkit#10]
---
title: 
subtitle: 
description: 
slug: 
tags: []
category: blog

date: 2023-08-31
canonical: https://blog.trangml.com
---

Questions:

- What is the Jacobian? What is the Hessian? In machine learning, what are they useful for?
	- The Jacobian is a matrix of the first-order partial derivatives of a vector-valued function. In robotics, the Jacobian is crucial for understanding the relationship between the velocity of robot joints and the velocity of the end effector. In machine learning, it's used for sensitivity analysis and to understand how changes in inputs affect outputs.
	- Hessian: The Hessian matrix is a square matrix of second-order partial derivatives of a scalar-valued function. It's used in optimization problems to find local maxima and minima, which is essential in training machine learning models, especially for understanding the curvature of the loss landscape.

- What is matrix decomposition and what are some algorithms that do it?
	- Matrix decomposition involves breaking down a matrix into a product of matrices to simplify certain matrix computations and analysis. Common algorithms include:
	-     LU Decomposition: Factorizes a matrix into Lower and Upper triangular matrices.
	-     QR Decomposition: Decomposes a matrix into an orthogonal matrix (Q) and an upper triangular matrix (R).
	-     Eigen Decomposition: Factorizes a matrix into its eigenvectors and eigenvalues.
	-     Singular Value Decomposition (SVD): Breaks down a matrix into singular vectors and singular values.
	- These decompositions are used in solving linear equations, optimization problems, and principal component analysis (PCA), among others, in robotics and machine learning.

- What is a transformer model? Can I implement it from scratch?
	- A Transformer model is a deep learning model that uses self-attention mechanisms to process sequential data, such as text or time-series data, without relying on recurrence. 
	- Transformers are quadratic in sequence width, which is why they are slower than RNNs

- What is a generative model?
	- A generative model learns the relationship between input and output distributions, and can generate new data based on what it's trained on. Generative Adversarial Networks, Variational Autoencoders, etc.
	
- How do you solve the vanishing gradients in LSTM?
	- LSTMs introduce a series of games (input, forget, and output), as well as internal cell states. The gates allow LSTMs to selectively remember or forget info.
 
- Know the fundamentals of RNN, LSTM, Transformer, GNN, GAN, CNN
	-    Recurrent Neural Networks (RNN): Designed to handle sequential data by maintaining a hidden state that captures information from previous inputs.
    - Long Short-Term Memory (LSTM): An advanced RNN that uses gating mechanisms to control the flow of information, addressing the vanishing gradient problem.
    - Transformer: Uses self-attention mechanisms to weigh the significance of different parts of the input data differently, excelling in handling sequential data without the sequential computation of RNNs.
    - Graph Neural Networks (GNN): Processes graph-structured data, updating node representations based on their neighbors, ideal for social networks, molecule structures, etc.
    - Generative Adversarial Networks (GAN): Consists of a generator and discriminator competing against each other, useful for generating realistic images, videos, etc.
    - Convolutional Neural Networks (CNN): Specialized for processing structured grid data like images, using convolutional layers to capture spatial hierarchies.

- What is a contrastive representation learning algorithm?
	- Contrastive representation learning algorithms learn to encode similar data points closer together and dissimilar ones farther apart in the representation space. This is achieved by using contrastive loss functions, which compare pairs or groups of examples to learn discriminative features. These algorithms are essential for unsupervised learning, semi-supervised learning, and self-supervised learning tasks, improving performance in downstream tasks like classification without requiring labeled data.
	
- How do classical robot navigation and path planning algorithms work?
	- Classical robot navigation involves making a grid or graph based map, then using algorithms like A*, Dijkstra's Algorithm, or Rapidly-exploring Random Trees (RRT) to find a path in the grid.

- How do classical robot grippers work?
	- Model Predictive Control algorithms. Inverse Kinematics. 
	- Inverse kinematics is the process of calcuating the joint parameters needed to place the end effector of a robot at a desired position and orientation in space. 
	- Model Predictive control uses a model of the robots dynamics to predict its future state and optimize control inputs to achieve desired outcomes. It's good because it's adaptive and will constantly change to catch up with the challenge.
	
- What are some normalization algorithms and what do they do?
	- Normalization algorithms in machine learning and deep learning include Batch Normalization, Layer Normalization, Instance Normalization, and Group Normalization. These techniques adjust and scale the inputs or the activations of a network to stabilize and accelerate the training process, reduce the sensitivity to initialization, and help combat the vanishing/exploding gradient problems.
- What is a hard problem you've had to solve?
- 
- What are some challenges you've overcome?
- 
- What is something you learned over the last year?
- 
- What is Bayes rule?
	- Bayes' rule is a fundamental theorem in probability theory and statistics that describes the probability of an event, based on prior knowledge of conditions that might be related to the event.  In robotics and machine learning, Bayes' rule is used for Bayesian inference, filtering, and decision making under uncertainty.

Takeaways:
- review linear algebra
- review classical robotics
- review different machine learning models and learning frameworks

Ideas on how I can improve myself
- take a crash course|brilliant course on linear algebra and statistics
- read a fundamental paper each week
- implement transformer from scratch
- create educational content
